File: go/neetcode/array_hashing/242.Valid_Anagram.go
package array_hashing

import (
	"fmt"
	"sort"
	"strings"
)

func isAnagram(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	var mapping_t map[rune]int = make(map[rune]int)
	for _, str := range t {
		mapping_t[str] += 1
	}

	for _, str := range s {
		if mapping_t[str] == 0 {
			return false
		}
		mapping_t[str] -= 1
	}
	return true
}

func isAnagram2(s string, t string) bool {
	if len(s) != len(t) {
		return false
	}

	sn := strings.Split(s, "")
	sort.Strings(sn)
	tn := strings.Split(t, "")
	sort.Strings(tn)

	for pos, str := range tn {
		if sn[pos] != str {
			return false
		}
	}
	return true
}

func Execute_isAnagram() {
	s := "anagram"
	// s = "a"
	t := "nagaram"
	// t = "ab"
	ans := isAnagram2(s, t)
	fmt.Println(ans)
}

File: py/242_Valid_Anagram.py
def isAnagram(s: str, t: str) -> bool:
    d = dict()
    count = len(s)
    if len(s) != len(t):
        return False

    for num in range(count):
        d[s[num]] = d.get(s[num], 0) + 1
        d[t[num]] = d.get(t[num], 0) - 1

    for key, val in d.items():
        if val != 0:
            return False
    return True
